chunky/src/java/se/llbit/chunky/world/ImposterCubicChunk.java

```diff
@@ -0,0 +1,200 @@
+package se.llbit.chunky.world;
+
+import se.llbit.chunky.block.legacy.LegacyBlocks;
+import se.llbit.chunky.chunk.BlockPalette;
+import se.llbit.chunky.chunk.ChunkData;
+import se.llbit.chunky.chunk.EmptyChunkData;
+import se.llbit.chunky.chunk.GenericChunkData;
+import se.llbit.chunky.map.IconLayer;
+import se.llbit.chunky.map.SurfaceLayer;
+import se.llbit.chunky.world.region.ImposterCubicRegion;
+import se.llbit.math.QuickMath;
+import se.llbit.nbt.*;
+import se.llbit.util.BitBuffer;
+import se.llbit.util.Mutable;
+import se.llbit.util.NotNull;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+public class ImposterCubicChunk extends Chunk {
+  private final CubicWorld world;
+
+  public ImposterCubicChunk(ChunkPosition pos, World world) {
+    super(pos, world);
+    assert world instanceof CubicWorld;
+    this.world = (CubicWorld) world;
+
+    assert world.getVersionId() <= 1343;
+    version = "1.12";
+  }
+
+  private Map<Integer, Map<String, Tag>> getCubeTags(Set<String> request) {
+    Mutable<Integer> timestamp = new Mutable<>(dataTimestamp);
+    ImposterCubicRegion region = (ImposterCubicRegion) world.getRegion(position.getRegionPosition());
+    Map<Integer, Map<String, Tag>> cubeTagsInColumn = region.getCubeTagsInColumn(position, request, timestamp);
+    dataTimestamp = timestamp.get();
+    return cubeTagsInColumn;
+  }
+
+  /**
+   * Parse the chunk from the region file and render the current
+   * layer, surface and cave maps.
+   * @return whether the input chunkdata was modified
+   */
+  public synchronized boolean loadChunk(ChunkData chunkData, int yMin, int yMax) {
+    if (!shouldReloadChunk()) {
+      return false;
+    }
+
+    Set<String> request = new HashSet<>();
+    request.add(Chunk.DATAVERSION);
+    request.add(Chunk.LEVEL_SECTIONS);
+    request.add(Chunk.LEVEL_BIOMES);
+    Map<Integer, Map<String, Tag>> data = getCubeTags(request);
```

Comment:
[Int2ObjectMap](https://fastutil.di.unimi.it/docs/it/unimi/dsi/fastutil/ints/Int2ObjectMap.html)?